streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java

```diff
@@ -0,0 +1,478 @@
+                .filter(o -> !unavailableClients.contains(o))
+                .collect(Collectors.toSet());
+        }
+
+        public double clientLoad(final ProcessId processId) {
+            final int capacity = clients.get(processId).numProcessingThreads();
+            final double totalTaskCount = newAssignments.getOrDefault(processId, new HashSet<>()).size();
+            return totalTaskCount / capacity;
+        }
+
+        public ProcessId findLeastLoadedClient(final TaskId taskId, final Set<ProcessId> clientIds) {
+            ProcessId leastLoaded = null;
+            for (final ProcessId processId : clientIds) {
+                final double thisClientLoad = clientLoad(processId);
+                if (thisClientLoad == 0) {
+                    return processId;
+                }
+
+                if (leastLoaded == null || thisClientLoad < clientLoad(leastLoaded)) {
+                    final Set<TaskId> assignedTasks = newAssignments.getOrDefault(processId, new HashSet<>())
+                        .stream().map(AssignedTask::id).collect(Collectors.toSet());
+                    if (taskPairs.hasNewPair(taskId, assignedTasks)) {
+                        leastLoaded = processId;
+                    }
+                }
+            }
+
+            if (leastLoaded != null) {
+                return leastLoaded;
+            }
+
+            for (final ProcessId processId : clientIds) {
+                final double thisClientLoad = clientLoad(processId);
+
+                if (leastLoaded == null || thisClientLoad < clientLoad(leastLoaded)) {
+                    leastLoaded = processId;
+                }
+            }
+
+            return leastLoaded;
+        }
+
+        public ProcessId findLeastLoadedClientWithPreviousActiveOrStandbyTask(final TaskId taskId,
+                                                                     final Set<ProcessId> clientsWithin) {
+            final ProcessId previous = previousActiveAssignment.get(taskId);
+            if (previous != null && clientsWithin.contains(previous)) {
+                return previous;
+            }
+            return findLeastLoadedClientWithPreviousStandbyTask(taskId, clientsWithin);
+        }
+
+        public ProcessId findLeastLoadedClientWithPreviousStandbyTask(final TaskId taskId,
+                                                                      final Set<ProcessId> clientsWithin) {
+            final Set<ProcessId> ids = previousStandbyAssignment.getOrDefault(taskId, new HashSet<>());
+            final HashSet<ProcessId> constrainTo = new HashSet<>(ids);
+            constrainTo.retainAll(clientsWithin);
+            return findLeastLoadedClient(taskId, constrainTo);
+        }
+
+        public boolean shouldBalanceLoad(final ProcessId client) {
+            if (clientLoad(client) < 1) {
+                return false;
+            }
+
+            final double thisClientLoad = clientLoad(client);
```

Comment:
can we declare this above so we can use it in the `if (clientLoad(client) < 1) {` line as well?